py/desispec/resolution.py

```diff
@@ -33,18 +34,27 @@ class Resolution(scipy.sparse.dia_matrix):
     """
     def __init__(self,data):
 
-        if scipy.sparse.isspmatrix_dia(data) and np.array_equal(np.sort(data.offsets)[::-1],self.offsets):
+        ### if scipy.sparse.isspmatrix_dia(data) and np.array_equal(np.sort(data.offsets)[::-1],self.offsets):
+        if scipy.sparse.isspmatrix_dia(data):
             # Input is already in DIA format with the required diagonals.
             # We just need to put the diagonals in the correct order.
             diag_order = np.argsort(data.offsets)[::-1]
+            ndiag = len(data.offsets)
+            assert ndiag%2 == 1, "Number of diagonals ({}) should be odd".format(ndiag)
+            self.offsets = np.arange(ndiag//2,-(ndiag//2)-1,-1)
+            if not np.array_equal(data.offsets[diag_order], self.offsets):
+                raise ValueError('Offsets of input matrix are non-contiguous or non-symmetric')
             scipy.sparse.dia_matrix.__init__(self,(data.data[diag_order],self.offsets),data.shape)
 
         elif isinstance(data,np.ndarray) and len(data.shape) == 2:
             n1,n2 = data.shape
-            if n1 == num_diagonals and n2 > n1:
+            if n2 > n1:
+                ndiag = data.shape[0]
+                self.offsets = np.arange(ndiag//2,-(ndiag//2)-1,-1)
```

Comment:
Well, well.  A bug in my unit test. :)

    #- An even number if diagonals is not allowed
    try:
        ndiag = 10
        R7 = Resolution(np.ones((ndiag, n)))
        raise ValueError('Incorrectly created Resolution with even number of diagonals')
    except ValueError, err:
        #- it correctly raised an error, so pass
        pass

The code was accidentally correctly failing to create a Resolution matrix with an even number of diagonals because the number of offsets ended up as odd anyway and dia_matrix.__init__ complained that the number of diagonals didn't match the number of offsets.  But even if it had (incorrectly) succeeded my test made matters worse by raising a ValueError -- the same type of error it was looking to catch so it would have passed anyway.

So thanks for the code review catching that I wasn't checking for ndiag odd here.  I'll update and submit a new pull request since I already merged this one (oops).